### DISCUSSION TITLE
Can I add a custom plugin to Kong Free Plan not Enterprise?

### DISCUSSION THREAD

**Discussion by tvluongict** [Kong/kong#9030] [Help] [answered]
Hi everyone.
I'm using a Kong - free plan
Can I add a custom plugin into my Kong (free plan)?
I'm following this article: https://konghq.com/blog/custom-authentication-and-authorization-framework-with-kong
But the custom plugin is not showing

**Comment by Tieske** [comment, marked as answer]
yes you can. Please see https://docs.konghq.com/gateway/2.8.x/plugin-development/distribution/

**Reply by tvluongict** [reply]
Hi @Tieske 
Thank you so much for your support.
My kong (free plan) is running inside a docker container.
And I built custom-auth plugin based on this article: https://konghq.com/blog/custom-authentication-and-authorization-framework-with-kong
Do you know why it doesn't show in my list plugin?
I think the reason is that I am using kong free plan not enterprise. Am I right?

**Reply by naolaregaqena** [reply]
are you using konga as your admin ui. if so, it could be because konga populates it's plugin list from the `available_on_server` property in the admin api. i have faced the same issue and also filed and issue. from what i saw in the master branch it is going to load it's available plugin list from the loaded list in the next kong major release. but first make sure it is showing in `loaded_plugins` in your admin api root endpoint `http://{kong admin host}:8001`.

**Reply by tvluongict** [reply]
Hi @naol-arega ,
Thank you for your support first
I checked and see that:
- it is showing in the list **loaded_plugins**
- it is now showing in the list: **available_on_server**


What should I do next?
This is my kong config:

"loaded_plugins": {
            "application-registration": true,
            "oauth2-introspection": true,
            "proxy-cache-advanced": true,
            "openid-connect": true,
            "forward-proxy": true,
            "canary": true,
            "request-transformer-advanced": true,
            **"custom-auth": true,**
            "rate-limiting-advanced": true,
            "ldap-auth-advanced": true,
            "statsd-advanced": true,
            "route-by-header": true,
            "jwt-signer": true,
            "vault-auth": true,
            "collector": true,
            "request-validator": true,
            "mtls-auth": true,
            "correlation-id": true,
            "cors": true,
            "degraphql": true,
            "route-transformer-advanced": true,
            "udp-log": true,
            "file-log": true,
            "http-log": true,
            "key-auth": true,
            "hmac-auth": true,
            "basic-auth": true,
            "ip-restriction": true,
            "request-transformer": true,
            "response-transformer": true,
            "request-size-limiting": true,
            "rate-limiting": true,
            "response-ratelimiting": true,
            "syslog": true,
            "loggly": true,
            "datadog": true,
            "ldap-auth": true,
            "statsd": true,
            "bot-detection": true,
            "aws-lambda": true,
            "request-termination": true,
            "prometheus": true,
            "proxy-cache": true,
            "session": true,
            "acme": true,
            "grpc-gateway": true,
            "grpc-web": true,
            "pre-function": true,
            "post-function": true,
            "azure-functions": true,
            "zipkin": true,
            "jwt": true,
            "response-transformer-advanced": true,
            "oauth2": true,
            "tcp-log": true,
            "acl": true,
            "graphql-rate-limiting-advanced": true,
            "jq": true,
            "opa": true,
            "mocking": true,
            "upstream-timeout": true,
            "key-auth-enc": true,
            "exit-transformer": true,
            "kafka-upstream": true,
            "graphql-proxy-cache-advanced": true,
            "hide-token": true,
            "kafka-log": true
        },


"available_on_server": {
            "jwt": true,
            "acl": true,
            "correlation-id": true,
            "cors": true,
            "oauth2": true,
            "tcp-log": true,
            "udp-log": true,
            "file-log": true,
            "http-log": true,
            "key-auth": true,
            "hmac-auth": true,
            "basic-auth": true,
            "ip-restriction": true,
            "request-transformer": true,
            "response-transformer": true,
            "request-size-limiting": true,
            "rate-limiting": true,
            "response-ratelimiting": true,
            "syslog": true,
            "loggly": true,
            "datadog": true,
            "ldap-auth": true,
            "statsd": true,
            "bot-detection": true,
            "aws-lambda": true,
            "request-termination": true,
            "prometheus": true,
            "proxy-cache": true,
            "session": true,
            "acme": true,
            "grpc-gateway": true,
            "grpc-web": true,
            "pre-function": true,
            "post-function": true,
            "azure-functions": true,
            "zipkin": true
        },

**Reply by naolaregaqena** [reply]
how did you install your custom plugin, for me i had to build a custom image based on `kong:2.8.1` which i installed my custom plugin luarock. and also the enterprise edition `kong/kong-gataway:2.8-alpine` docker image has the issue with showing plugins but the `kong:2.8.1` seems to work as intended. you might encounter some issue with `openssl` if you are using `postgres:14.x`. you have to use `postgres:13.x` for it to work.

**Reply by tvluongict** [reply]
1) how did you install your custom plugin
- I copy custom-auth to folder: /usr/local/share/lua/5.1/kong/plugins (inside kong container)

2) my kong version: 2.8.1.1-enterprise-edition

3) my postgres version: postgres:9.6

**Reply by naolaregaqena** [reply]
i created a `Dockerfile` based on `kong:2.8.1`. i installed it with `luarocks install {custom plugin}` command. then added a `kong.conf` file to `/etc/kong/kong.conf` in the image with the line `plugins = bundled,{custom plugin}` in it. based on this custom image i bootstrapped and run kong.

**Reply by Tieske** [reply]
It is listed, so it is available and can be used. Check the docs on how to configure a plugin.

**Reply by tvluongict** [reply]
Thanks @Tieske, @naol-arega for your support.
I don't know why the plugin is not showing in KONGA plugin page or KONG Manger Plugin page, but I can use it by using command line.
